### PR TITLE
feat: added initialData overloads for useQuery

### LINKED ISSUE #2996: feat: add overloads to useQuery
### Describe the feature you'd like to request

Ref: https://discord.com/channels/867764511159091230/1030047679977357312

tanstack/query has [some overloads](https://github.com/TanStack/query/blob/main/packages/react-query/src/useQuery.ts) when passing `initialData` as options, making the type defined. This is missing in tRPC.

image

### Describe the solution you'd like to see

add overloads that uses the `DefinedUseQueryResult`: https://github.com/TanStack/query/blob/afbd6c73776ad2d68b6d5b64ce9888c85c745a90/packages/react-query/src/types.ts#L76

### Desribe alternate solutions

not adding

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Closes #2996

## 🎯 Changes

adds function overloads to two forms of `useQuery`:

* `trpc.useQuery`: with an inline function overload
* `ctx.proxy.post.byId.useQuery` & the like: with an interface containing multiple call signatures

Both overloads indicate that if `opts.initialData` is defined, the result should have a defined `data`.

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/react-query/src/shared/hooks/deprecated/createHooksInternal.tsx [hunk 1]
```diff
@@ -291,6 +293,34 @@ function __createHooksInternal<
...
+      TQueryValues[TPath]['input'],
+      TQueryFnData,
+      TData,
+      TError
+    >,
+  ): UseTRPCQueryResult<TData, TError>;
+  function useQuery<
+    TPath extends keyof TQueryValues & string,
+    TQueryFnData,
+    TData,
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/deprecated/createHooksInternal.tsx:323]: This is v9 which is deprecated and will be deleted (see #3496)

Better to leave it untouched.

```suggestion
```

**Review by KATT — CHANGES_REQUESTED** [review]
Love it, but let's not add this in the v9 (any `/interop/*` or `/deprecated/*` - see #3496, we've already deleted a bunch of these files in #3494)

File: packages/react-query/src/shared/hooks/deprecated/createHooksInternal.tsx (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/react-query/src/shared/hooks/deprecated/createHooksInternal.tsx:323]: Nice!!

File: packages/react-query/src/shared/hooks/types.ts [hunk 2]
```diff
@@ -49,6 +51,16 @@ export interface UseTRPCQueryOptions<TPath, TInput, TOutput, TData, TError>
   extends UseQueryOptions<TOutput, TError, TData, [TPath, TInput]>,
     TRPCUseQueryBaseOptions {}
 
+export interface DefinedUseTRPCQueryOptions<
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/types.ts:54]:
```suggestion
/** @internal **/
export interface DefinedUseTRPCQueryOptions<
```

**Review by KATT — DISMISSED** [review]
I'd love @juliusmarminge to look at this cause he raised the original issue 👀

**Review by KATT — DISMISSED** [review]

**Comment by KATT** [comment]
cc @juliusmarminge smash the merge button if you like it

**Comment by juliusmarminge** [comment]
> cc @juliusmarminge smash the merge button if you like it

Will look in a bit

**Review by juliusmarminge — APPROVED** [review]
Nice job!

I added back the original regression tests that verifies that "all" methods return the same as tanstack (destructuring, initialData, fallbackData).

**Review by KATT — APPROVED** [review]

**Pull request merged**